File: native/shim/proto.go
// Wire protocol between the BEAM (host) and the shim.
//
// Every packet is `<<len::32-big, tag::8, payload::binary>>` where len covers
// the tag byte plus the payload — exactly what an Erlang port opened with
// `{:packet, 4}` produces and expects. The framing and the demand-driven
// read model are adapted from odu (github.com/akash-akya/ex_cmd, MIT,
// Copyright Akash Hiremath), itself based on goon (github.com/alco/goon, MIT,
// Copyright Alexei Sholik). See NOTICE at the repository root.
package main

import (
	"encoding/binary"
	"errors"
	"fmt"
	"io"
)

// ProtocolVersion must match Longx.Shim.Proto on the Elixir side.
const ProtocolVersion = "3"

// Packets sent by the host to the shim.
const (
	TagInput       uint8 = 1  // payload: bytes for the child's stdin
	TagCloseInput  uint8 = 2  // close the child's stdin
	TagSendOutput  uint8 = 3  // payload: max::32 — read at most `max` bytes from stdout once
	TagSendStderr  uint8 = 4  // payload: max::32 — same for stderr
	TagCloseOutput uint8 = 5  // stop reading stdout
	TagCloseStderr uint8 = 6  // stop reading stderr
	TagKill        uint8 = 7  // payload: grace_ms::32 — terminate the tree, SIGKILL after grace
	TagSignal      uint8 = 8  // payload: signum::32 — forward a signal to the child
	TagCommandEnv  uint8 = 9  // payload: [len::16, "K=V"]* — must be the first packet
	TagSendStats   uint8 = 10 // answer with one Stats packet for the child's process tree
)

// Packets sent by the shim to the host.
const (
	TagPid        uint8 = 16 // payload: pid::32
	TagOutput     uint8 = 17 // payload: stdout bytes (answer to SendOutput)
	TagOutputEOF  uint8 = 18 // stdout reached EOF or was closed
	TagStderr     uint8 = 19 // payload: stderr bytes (answer to SendStderr)
	TagStderrEOF  uint8 = 20
	TagExitStatus uint8 = 21 // payload: status::32-signed
	TagStartError uint8 = 22 // payload: reason string; the child never started
	TagSendInput  uint8 = 23 // the child is ready for one more Input packet
	TagStats      uint8 = 24 // payload: JSON {"processes","rss_bytes","cpu_ms"} (answer to SendStats)
)

// MaxPayload is the largest payload that fits in a single packet. Kept at
// 64 KiB minus framing so one packet always fits one Erlang port message.
const MaxPayload = (1 << 16) - 5

const headerSize = 4 + 1

var (
	errPayloadTooLarge = errors.New("shim: payload exceeds MaxPayload")
	errBadLength       = errors.New("shim: invalid packet length")
	errShortField      = errors.New("shim: field too short")
)

// Packet is one decoded frame.
type Packet struct {
	Tag  uint8
	Data []byte
}

// readPacket reads one frame. It returns io.EOF only on a clean EOF at a
// frame boundary; a truncated frame is io.ErrUnexpectedEOF.
func readPacket(r io.Reader) (Packet, error) {
	var head [headerSize]byte

	if _, err := io.ReadFull(r, head[:4]); err != nil {
		if err == io.ErrUnexpectedEOF {
			return Packet{}, err
		}
		return Packet{}, io.EOF
	}

	length := binary.BigEndian.Uint32(head[:4])
	if length < 1 || length-1 > MaxPayload {
		return Packet{}, errBadLength
	}

	if _, err := io.ReadFull(r, head[4:5]); err != nil {
		return Packet{}, io.ErrUnexpectedEOF
	}

	data := make([]byte, length-1)
	if _, err := io.ReadFull(r, data); err != nil {
		return Packet{}, io.ErrUnexpectedEOF
	}

	return Packet{Tag: head[4], Data: data}, nil
}

// writePacket writes one frame with a single Write call so concurrent writers
// serialised by a mutex never interleave frames.
func writePacket(w io.Writer, tag uint8, data []byte) error {
	if len(data) > MaxPayload {
		return errPayloadTooLarge
	}

	frame := make([]byte, headerSize+len(data))
	binary.BigEndian.PutUint32(frame[:4], uint32(len(data)+1))
	frame[4] = tag
	copy(frame[5:], data)

	_, err := w.Write(frame)
	return err
}

// decodeEnv parses the CommandEnv payload: repeated `<<len::16, "KEY=VALUE">>`.
func decodeEnv(data []byte) ([]string, error) {
	var env []string
	for i := 0; i < len(data); {
		if i+2 > len(data) {
			return nil, fmt.Errorf("shim: truncated env length at offset %d", i)
		}
		n := int(binary.BigEndian.Uint16(data[i : i+2]))
		i += 2
		if i+n > len(data) {
			return nil, fmt.Errorf("shim: truncated env entry at offset %d", i)
		}
		env = append(env, string(data[i:i+n]))
		i += n
	}
	return env, nil
}

func encodeUint32(n uint32) []byte {
	b := make([]byte, 4)
	binary.BigEndian.PutUint32(b, n)
	return b
}

func decodeUint32(b []byte) (uint32, error) {
	if len(b) != 4 {
		return 0, errShortField
	}
	return binary.BigEndian.Uint32(b), nil
}

File: lib/longx/shim/proto.ex
defmodule Longx.Shim.Proto do
  @moduledoc """
  Wire protocol between `Longx.Shim` and the Go shim in `native/shim`.

  The port is opened with `{:packet, 4}`, so each message here is just
  `<<tag::8, payload::binary>>`; the BEAM adds the length prefix. Tags and
  payload layouts mirror `native/shim/proto.go` — keep both in sync and bump
  `version/0` when the format changes.
  """

  @version "4"

  # host -> shim
  @input 1
  @close_input 2
  @send_output 3
  @send_stderr 4
  @close_output 5
  @close_stderr 6
  @kill 7
  @signal 8
  @env 9
  @send_stats 10

  # shim -> host
  @pid 16
  @output 17
  @output_eof 18
  @stderr 19
  @stderr_eof 20
  @exit_status 21
  @start_error 22
  @send_input 23
  @stats 24

  # 4 byte length prefix + 1 byte tag leaves this much payload in a 64 KiB packet
  @max_chunk 64 * 1024 - 5
  @max_env_entry 0xFFFF

  @type command ::
          :close_input
          | :close_output
          | :close_stderr
          | :send_stats
          | {:input, binary}
          | {:send_output, pos_integer}
          | {:send_stderr, pos_integer}
          | {:kill, non_neg_integer}
          | {:signal, non_neg_integer}
          | {:env, [{String.t(), String.t()}]}

  @type event ::
          {:pid, pos_integer}
          | {:output, binary}
          | {:stderr, binary}
          | :output_eof
          | :stderr_eof
          | {:exit_status, integer}
          | {:start_error, String.t()}
          | :send_input
          | {:stats, stats}
          | {:unknown, byte, binary}

  @typedoc "The child's whole process tree at one instant."
  @type stats :: %{
          processes: non_neg_integer,
          rss_bytes: non_neg_integer,
          cpu_ms: non_neg_integer
        }

  @doc "Protocol version the shim must report from `shim -v`."
  @spec version() :: String.t()
  def version, do: @version

  @doc "Largest `:input` payload / `:send_output` request that fits one packet."
  @spec max_chunk() :: pos_integer
  def max_chunk, do: @max_chunk

  @spec encode(:close_input | :close_output | :close_stderr | :send_stats) :: binary
  def encode(:close_input), do: <<@close_input>>
  def encode(:close_output), do: <<@close_output>>
  def encode(:close_stderr), do: <<@close_stderr>>
  def encode(:send_stats), do: <<@send_stats>>

  @spec encode(atom, term) :: binary
  def encode(:input, data) when is_binary(data) and byte_size(data) <= @max_chunk,
    do: <<@input, data::binary>>

  def encode(:send_output, max) when is_integer(max) and max > 0 and max <= @max_chunk,
    do: <<@send_output, max::unsigned-big-32>>

  def encode(:send_stderr, max) when is_integer(max) and max > 0 and max <= @max_chunk,
    do: <<@send_stderr, max::unsigned-big-32>>

  def encode(:kill, grace_ms) when is_integer(grace_ms) and grace_ms >= 0,
    do: <<@kill, grace_ms::unsigned-big-32>>

  def encode(:signal, signum) when is_integer(signum) and signum >= 0,
    do: <<@signal, signum::unsigned-big-32>>

  def encode(:env, env) when is_list(env) do
    entries =
      Enum.map(env, fn {key, value} ->
        entry = to_string(key) <> "=" <> to_string(value)

        if byte_size(entry) > @max_env_entry do
          raise ArgumentError, "env entry #{inspect(key)} exceeds #{@max_env_entry} bytes"
        end

        <<byte_size(entry)::unsigned-big-16, entry::binary>>
      end)

    IO.iodata_to_binary([@env | entries])
  end

  @spec decode(binary) :: event
  def decode(<<@pid, pid::unsigned-big-32>>), do: {:pid, pid}
  def decode(<<@output, data::binary>>), do: {:output, data}
  def decode(<<@output_eof>>), do: :output_eof
  def decode(<<@stderr, data::binary>>), do: {:stderr, data}
  def decode(<<@stderr_eof>>), do: :stderr_eof
  def decode(<<@exit_status, status::signed-big-32>>), do: {:exit_status, status}
  def decode(<<@start_error, reason::binary>>), do: {:start_error, reason}
  def decode(<<@send_input>>), do: :send_input

  def decode(<<@stats, json::binary>>) do
    case Jason.decode(json) do
      {:ok, %{"processes" => p, "rss_bytes" => rss, "cpu_ms" => cpu}}
      when is_integer(p) and is_integer(rss) and is_integer(cpu) ->
        {:stats, %{processes: p, rss_bytes: rss, cpu_ms: cpu}}

      _ ->
        {:unknown, @stats, json}
    end
  end

  def decode(<<tag, rest::binary>>), do: {:unknown, tag, rest}
end
